File: app/src/main/java/com/arkapp/partyplanner/ui/venueLocation/LocationAdapter.java
package com.arkapp.partyplanner.ui.venueLocation;

import android.annotation.SuppressLint;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.RecyclerView.ViewHolder;

import com.arkapp.partyplanner.R;
import com.arkapp.partyplanner.data.models.Location;
import com.arkapp.partyplanner.data.models.PartyDetails;
import com.arkapp.partyplanner.data.repository.PrefRepository;
import com.arkapp.partyplanner.databinding.RvLocationBinding;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

import static com.arkapp.partyplanner.utils.GlideUtilsKt.loadImage;
import static com.arkapp.partyplanner.utils.ViewUtilsKt.hide;
import static com.arkapp.partyplanner.utils.ViewUtilsKt.show;

/**
 * Created by Abdul Rehman on 28-02-2020.
 * Contact email - abdulrehman0796@gmail.com
 */


/**
 * Recycler view adapter of location selection before the venue selection
 */
public class LocationAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private final PrefRepository prefRepository;
    private List<Location> locations;
    private ArrayList<Location> locationSelected = new ArrayList<>();


    public LocationAdapter(@NotNull List<Location> locations, @NotNull PrefRepository prefRepository) {
        super();
        this.locations = locations;
        this.prefRepository = prefRepository;
    }

    @NotNull
    @Override
    public ViewHolder onCreateViewHolder(@NotNull ViewGroup parent, int viewType) {
        return (ViewHolder) (new LocationViewHolder(
                (RvLocationBinding) DataBindingUtil
                        .inflate(LayoutInflater.from(
                                parent.getContext()),
                                R.layout.rv_location,
                                parent,
                                false)));
    }

    @SuppressLint({"ClickableViewAccessibility", "SetTextI18n"})
    @Override
    public void onBindViewHolder(@NotNull ViewHolder holder, int position) {
        RvLocationBinding binding = ((LocationViewHolder) holder).getViewBinding();

        Location location = locations.get(position);
        binding.title.setText(location.getName());
        loadImage(binding.icon, location.getResId());

        if (locationSelected.contains(location))
            show(binding.tick);

        //Rotating the compass icon in different location
        switch (position) {
            case 0:
                binding.icon.setRotation(-45f);
                break;
            case 2:
                binding.icon.setRotation(135f);
                break;
            case 3:
                binding.icon.setRotation(-135f);
                break;
            case 4:
                binding.icon.setRotation(45f);
                break;
        }

        //Storing the selected location in the shared preferences
        binding.parent.setOnClickListener(v -> {
            if (binding.tick.getVisibility() == View.VISIBLE) {
                locationSelected.remove(location);
                hide(binding.tick);

                PartyDetails details = prefRepository.getCurrentPartyDetails();
                details.getLocations().remove(location.getName());
                prefRepository.setCurrentPartyDetails(details);

            } else {
                locationSelected.add(location);
                show(binding.tick);

                PartyDetails details = prefRepository.getCurrentPartyDetails();
                details.getLocations().add(location.getName());
                prefRepository.setCurrentPartyDetails(details);
            }
        });
    }

    public int getItemCount() {
        return this.locations.size();
    }

    public long getItemId(int position) {
        return (long) (locations.get(position)).hashCode();
    }
}

File: app/src/main/java/com/arkapp/partyplanner/ui/venueLocation/LocationAdapter.kt
package com.arkapp.partyplanner.ui.venueLocation

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.core.view.isVisible
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.arkapp.partyplanner.R
import com.arkapp.partyplanner.data.models.Location
import com.arkapp.partyplanner.data.repository.PrefRepository
import com.arkapp.partyplanner.utils.hide
import com.arkapp.partyplanner.utils.loadImage
import com.arkapp.partyplanner.utils.show

/**
 * Created by Abdul Rehman on 28-02-2020.
 * Contact email - abdulrehman0796@gmail.com
 */

class LocationAdapter(private val locations: List<Location>,
                      private val prefRepository: PrefRepository) :
    RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    var locationSelected = ArrayList<Location>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return LocationViewHolder(
            DataBindingUtil.inflate(
                LayoutInflater.from(parent.context),
                R.layout.rv_location,
                parent,
                false
            )
        )
    }

    @SuppressLint("ClickableViewAccessibility", "SetTextI18n")
    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val binding = (holder as LocationViewHolder).viewBinding

        val location = locations[position]
        binding.title.text = location.name
        binding.icon.loadImage(location.resId)

        if (locationSelected.contains(location))
            binding.tick.show()

        when (position) {
            0 -> binding.icon.rotation = -45f
            1 -> binding.icon.rotation = 135f
            3 -> binding.icon.rotation = -135f
            4 -> binding.icon.rotation = 45f
        }

        binding.parent.setOnClickListener {
            if (binding.tick.isVisible) {
                locationSelected.remove(location)
                binding.tick.hide()

                val details = prefRepository.getCurrentPartyDetails()
                details.locations?.remove(location.name)
                prefRepository.setCurrentPartyDetails(details)

            } else {
                locationSelected.add(location)
                binding.tick.show()

                val details = prefRepository.getCurrentPartyDetails()
                details.locations?.add(location.name)
                prefRepository.setCurrentPartyDetails(details)
            }
        }
    }


    override fun getItemCount() = locations.size

    override fun getItemId(position: Int): Long {
        return locations[position].hashCode().toLong()
    }

}